packages/terra-dropdown-button/src/DropdownButton.jsx

```diff
@@ -57,29 +57,60 @@ class DropdownButton extends React.Component {
     this.handleDropdownRequestClose = this.handleDropdownRequestClose.bind(this);
     this.handleKeyDown = this.handleKeyDown.bind(this);
     this.handleKeyUp = this.handleKeyUp.bind(this);
+    this.setButtonNode = this.setButtonNode.bind(this);
+    this.getButtonNode = this.getButtonNode.bind(this);
+    this.setListNode = this.setListNode.bind(this);
+    this.state = { isOpen: false, isActive: false, openedViaKeyboard: false };
+  }
+
+  setListNode(element) {
+    this.dropdownList = element;
+  }
 
-    this.state = { isOpen: false, isActive: false };
+  setButtonNode(node) {
+    this.buttonNode = node;
   }
 
-  handleDropdownButtonClick() {
+  getButtonNode() {
+    return this.buttonNode;
+  }
+
+  handleDropdownButtonClick(event) {
+    if (this.state.isOpen) {
```

Comment:
Um, can these be clubbed together? 

```
this.setState(prevState => ({ isOpen: !prevState.isOpen, openedViaKeyboard: !prevState.isOpen }));
```

Reply:
setting openedViaKeyboard as batch state update will break functionality of setting focus on list item. Since state update will not happen before DropDownList is rendered condition in focus-trap takes previous value of openedViaKeyboard and sets focus on button even when dropdown is opened by keyboard ( this happens when dropdown is initially opened with mouse click and reverse when opened with keyboard initially ).

Reply:
I'm wondering if openedViaKeyboard should be a state variable at all. Calling setState for openedViaKeyboard alone doesn't appear to change any of the rendering of the component. This may be a case where the variable makes sense as a class variable and not a state variable.

```jsx
this.openedViaKeyboard = false;
```

Does making a state change to openedViaKeyboard have an effect if the component is not in an isOpen state?

```jsx
this.setState({ openedViaKeyboard: false }); // Calling this but never setting isOpen to true/false
```

Reply:
There seems to be some sync issues with openedViaKeyboard and isOpen when the state is updated multiple times during different events. If those can be reconciled this can stay as a state variable.